### PR TITLE
(OLD) [Feature] Support LoveDA dataset

### PR DESCRIPTION
**Pull request by Junjue-Wang**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Invited by [MengzhangLI](https://github.com/MengzhangLI), I would like to add [LoveDA](https://github.com/Junjue-Wang/LoveDA) dataset into mmsegmentation.

## Modification

The corresponding dataset api and  dataset_prepare.md has been updated.

## BC-breaking (Optional)

None.

## Use cases (Optional)

The users can easily use the LoveDA dataset and benchmarked segmentation models with MMSegmentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMDet3D.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: mmseg/datasets/loveda.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+    The ``img_suffix`` and ``seg_map_suffix`` are both fixed to '.png'.
+    """
+    CLASSES = ('background', 'building', 'road', 'water', 'barren', 'forest', 'agricultural')
+
+    PALETTE = [[255, 255, 255], [255, 0, 0], [255, 255, 0], [0, 0, 255],
+               [159, 129, 183], [0, 255, 0], [255, 195, 128]]
+
+    def __init__(self, **kwargs):
+        super(LoveDADataset, self).__init__(
+            img_suffix='.png', seg_map_suffix='.png', reduce_zero_label=True, **kwargs)
```
Comment (Junjun2016) [inline mmseg/datasets/loveda.py:20]: Is it necessary to use `reduce_zero_label` in this dataset?

File: docs/dataset_prepare.md [hunk 2]
```diff
@@ -253,3 +253,31 @@ Since we only support test models on this dataset, you may only download [the va
...
+# download
+mkdir loveda && cd loveda
+wget https://drive.google.com/drive/folders/1ibYV0qwn4yuuh068Rnc-w4tPi0U0c-ti?usp=sharing
+
+# unzip
+unzip \*.zip
+
+# Convert into segmentation splits
+mkdir -p img_dir/train img_dir/val img_dir/test ann_dir/train ann_dir/val
+mv Train\Rural\images_png\* img_dir\train
```
Comment (Junjun2016) [inline docs/dataset_prepare.md:273]:
```suggestion
mv Train/Rural/images_png/* img_dir/train
```

File: docs/dataset_prepare.md [hunk 3]
```diff
@@ -253,3 +253,31 @@ Since we only support test models on this dataset, you may only download [the va
...
+
+For LoveDA dataset, please run the following command to download and re-organize the dataset.
+
+```shell
+# download
+mkdir loveda && cd loveda
+wget https://drive.google.com/drive/folders/1ibYV0qwn4yuuh068Rnc-w4tPi0U0c-ti?usp=sharing
+
+# unzip
+unzip \*.zip
```
Comment (Junjun2016) [inline docs/dataset_prepare.md:269]: Should it be `unzip ./*.zip` or something else?

File: configs/_base_/datasets/loveda.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
+# dataset settings
+dataset_type = 'LoveDADataset'
+data_root = 'data/loveda'
+img_norm_cfg = dict(
+    mean=[123.675, 116.28, 103.53], std=[58.395, 57.12, 57.375], to_rgb=True)
+crop_size = (512, 512)
+train_pipeline = [
+    # dict(type='LoadImageFromFile'),
```
Comment (Junjun2016) [inline configs/_base_/datasets/loveda.py:8]:
```suggestion
    dict(type='LoadImageFromFile'),
```

File: configs/_base_/datasets/loveda.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    dict(type='LoadAnnotations', reduce_zero_label=True),
+    dict(type='RandomCrop', crop_size=crop_size, cat_max_ratio=0.75),
+    dict(type='RandomFlip', prob=0.5),
+    dict(type='Normalize', **img_norm_cfg),
+    dict(type='DefaultFormatBundle'),
+    dict(type='Collect', keys=['img', 'gt_semantic_seg']),
+]
+test_pipeline = [
+    # dict(type='LoadImageFromFile'),
```
Comment (Junjun2016) [inline configs/_base_/datasets/loveda.py:17]:
```suggestion
    dict(type='LoadImageFromFile'),
```

File: mmseg/datasets/loveda.py (hunk 1 above)
Comment (Junjue-Wang) [reply mmseg/datasets/loveda.py:20]: Yes, because in our masks, the ignore_index = 0. It refers to no-data value.

File: docs/dataset_prepare.md (hunk 3 above)
Comment (Junjue-Wang) [reply docs/dataset_prepare.md:269]: The '\' refers to the inversion symbol. This can help unzip all the zips.

**Review by Junjun2016 — COMMENTED** [review]
We need to check the inference and training performance between mmsegmentaion and the original paper implementation.

File: docs/dataset_prepare.md (hunk 2 above)
Comment (Junjue-Wang) [reply docs/dataset_prepare.md:273]: Thanks for your revision!

**Comment by Junjun2016** [comment]
Hi @Junjue-Wang 
Thanks for your contribution, nice feature!
Please fix the lint error.

**Comment by Junjue-Wang** [comment]
> We need to check the inference and training performance between mmsegmentaion and the original paper implementation.

The original paper reports the accuracies on the test set. However, the labels of the test set are private. If you really want to test. Please submit the predictions on this [Contest](https://competitions.codalab.org/competitions/35865) and you will get the scores smoothly. Or you can directly report the accuracies on the providied validation set.

**Comment by Junjue-Wang** [comment]
> Hi @Junjue-Wang Thanks for your contribution, nice feature! Please fix the lint error.

Thanks for your careful revision. Am I going to re-pull request after modification? Or you can directly help me fix the lint errors?

File: mmseg/datasets/loveda.py (hunk 1 above)
Comment (Junjue-Wang) [reply mmseg/datasets/loveda.py:20]: The class indexes (7 classes) are follows:
'background'-1  'building' -2  'road' - 3 'water' -4 'barren' - 5 'forest' -6 'agricultural' -7
So I think 0 needs to be reduced.

**Comment by MengzhangLI** [comment]
> > Hi @Junjue-Wang Thanks for your contribution, nice feature! Please fix the lint error.
> 
> Thanks for your careful revision. Am I going to re-pull request after modification? Or you can directly help me fix the lint errors?

For further participation of our open source codebase project, I suggest you follow these links [Chinese article](https://zhuanlan.zhihu.com/p/339080395) and [English document](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md), you need to use `pre-commit` to make codes of your PR as uniform style like codebase. 

Feel free to ask us if you have any problems.

Best,

**Comment by MengzhangLI** [comment]
Hi, @Junjue-Wang . Thanks for your nice help!

We would test your code and try to train several models of MMSegmentation on LoveDA next week. 

From your paper, it seems that HRNet is best model, maybe we can try some transformer models such as Swin Transformer and Segformer.

Best,

**Comment by Junjue-Wang** [comment]
> Hi, @Junjue-Wang . Thanks for your nice help!
> 
> We would test your code and try to train several models of MMSegmentation on LoveDA next week.
> 
> From your paper, it seems that HRNet is best model, maybe we can try some transformer models such as Swin Transformer and Segformer.
> 
> Best,

Yes, I wish you all the best on the test. If you encounter any problems, please feel free to contact me.

**Comment by Junjun2016** [comment]
Please fix the lint error.
You could run 
```shell
pre-commit run --all-files
```
before commit.

**Comment by Junjue-Wang** [comment]
> Please fix the lint error. You could run
> 
> ```shell
> pre-commit run --all-files
> ```
> 
> before commit.
I am not very familiar with the use of pre-commit, and encountered the following error. How should I adddress it?
image

**Comment by Junjun2016** [comment]
> > Please fix the lint error. You could run
> > ```shell
> > pre-commit run --all-files
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > before commit.
> > I am not very familiar with the use of pre-commit, and encountered the following error. How should I adddress it?
> > image

@MengzhangLI

**Comment by Junjun2016** [comment]
Hi @Junjue-Wang 
Please refer to https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md.

**Comment by Junjun2016** [comment]
@Junjue-Wang 
It seems that your pre-commit behaviors are not consistent with ours which leads to too many files being changed.
@MengzhangLI Could you please fix it?

File: docs_zh-CN/dataset_prepare.md [hunk 6]
```diff
@@ -195,37 +195,3 @@ python tools/convert_datasets/stare.py /path/to/stare-images.tar /path/to/labels
...
-mv Val/Urban/images_png/* img_dir/val
-mv Test/Rural/images_png/* img_dir/val
-mv Test/Urban/images_png/* img_dir/val
-mv Train/Rural/masks_png/* ann_dir/train
-mv Train/Urban/masks_png/* ann_dir/train
-mv Val/Rural/masks_png/* ann_dir/val
-mv Val/Urban/masks_png/* ann_dir/val
-```
-
-关于 LoveDA 的更多细节可以在[这里](https://github.com/Junjue-Wang/LoveDA)找到。
```
Comment (Junjun2016) [inline docs_zh-CN/dataset_prepare.md]: Why remove this description?

Comment (Junjue-Wang) [reply docs_zh-CN/dataset_prepare.md]: Sorry for this confusion. 
Because it is really difficult for me to install the pre-commit environment (Incompatible system, old version). So I manually revised the entire catalog with the mentioned tools in https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md.
Please help me check the lint errors in the last version. Thank you very much!

> @Junjue-Wang It seems that your pre-commit behaviors are not consistent with ours which leads to too many files being changed. @MengzhangLI Could you please fix it?

Comment (Junjue-Wang) [reply docs_zh-CN/dataset_prepare.md]: > Why remove this description?

This is my mistake, please reject this push.

Comment (Junjun2016) [reply docs_zh-CN/dataset_prepare.md]: You could push it again.

Comment (Junjue-Wang) [reply docs_zh-CN/dataset_prepare.md]: > You could push it again.

I have pushed the previous version again. Please check.

Comment (Junjun2016) [reply docs_zh-CN/dataset_prepare.md]: Great.

**Comment by MengzhangLI** [comment]
Update:

Hi, @Junjue-Wang 

Please grant authorization on me about your [forked MMSegmentation](https://github.com/Junjue-Wang/mmsegmentation/) followed [here](https://docs.github.com/cn/account-and-profile/setting-up-and-managing-your-github-user-account/managing-access-to-your-personal-repositories/inviting-collaborators-to-a-personal-repository). Thus I could push my modifications on your branch.

Here is some results on validation set:
image

FYI, the results of original paper:
image


TO DO:

- [ ] Getting results of models above on test set.

- [ ] Adding `format-only` in `./mmseg/datasets/loveda.py`, enabling users of MMSegmentation could directly generating standard prediction output of test set for evaluation on  [official codaLab site](https://competitions.codalab.org/competitions/35865#).

**Comment by Junjue-Wang** [comment]
> Update:
> 
> Hi, @Junjue-Wang
> 
> Please grant authorization on me about your [forked MMSegmentation](https://github.com/Junjue-Wang/mmsegmentation/) followed [here](https://docs.github.com/cn/account-and-profile/setting-up-and-managing-your-github-user-account/managing-access-to-your-personal-repositories/inviting-collaborators-to-a-personal-repository). Thus I could push my modifications on your branch.
> 
> Here is some results on validation set: image
> 
> FYI, the results of original paper: image
> 
> TO DO:
> 
> * [ ]  Getting results of models above on test set.
> * [ ]  Adding `format-only` in `./mmseg/datasets/loveda.py`, enabling users of MMSegmentation could directly generating standard prediction output of test set for evaluation on  [official codaLab site](https://competitions.codalab.org/competitions/35865#).

I have already granted the authorization to you. Please help me check the lint errors, Thank you!

**Comment by Junjun2016** [comment]
Too many files changed.
We may need to create a new PR.

File: mmseg/datasets/__init__.py [hunk 7]
```diff
@@ -9,16 +9,31 @@
...
+    'PascalContextDataset',
+    'PascalContextDataset59',
+    'ChaseDB1Dataset',
+    'DRIVEDataset',
+    'HRFDataset',
+    'STAREDataset',
+    'DarkZurichDataset',
+    'NightDrivingDataset',
+    'COCOStuffDataset',
+    'LoveDADataset',
```
Comment (Junjun2016) [inline mmseg/datasets/__init__.py:38]:
```suggestion
    'LoveDADataset'
```

Comment (Junjue-Wang) [reply mmseg/datasets/__init__.py:38]: @MengzhangLI What should we do? Create a new PR for what?

**Comment by MengzhangLI** [comment]
Hi, @Junjue-Wang 
This PR will be closed because too much changed files.
Could you please create a new pr thus I could fetch your branch and push to it?
